lib/raft/raft.hpp

```diff
@@ -0,0 +1,512 @@
+            bool success;
+            Index last_index;
+        } append_entries_reply;
+    };
+
+    Message()
+    {
+        // Set every field to zero
+        std::memset(this, 0, sizeof(*this));
+    }
+};
+
+template <typename StateMachine>
+class Peer {
+public:
+    Peer(NodeId id) : id(id)
+    {
+    }
+    virtual void send(const Message<StateMachine> &msg) = 0;
+    virtual ~Peer()
+    {
+    };
+
+    NodeId id;
+
+    // Those are per-peer information needed by the raft leader to send correct
+    // versions of the log to each peer
+    Index match_index;
+    Index next_index;
+};
+
+template <typename StateMachine>
+class State {
+public:
+    using Message = raft::Message<StateMachine>;
+    using Peer = raft::Peer<StateMachine>;
+    NodeId id;
+    Peer **peers;
+    int peer_count;
+    Term term;
+    int vote_count;
+    NodeId voted_for;
+    NodeState node_state;
+    int heartbeat_timer;
+    int election_timer;
+    // TODO: How many entries
+    Log<typename StateMachine::Operation, 10> log;
+    Index commit_index;
+
+    StateMachine &state_machine;
+
+    State(StateMachine &state_machine, NodeId id, Peer **peers, int peer_count) :
+        id(id), peers(peers), peer_count(peer_count), term(0), voted_for(0),
+        node_state(NodeState::Follower), heartbeat_timer(0),
+        election_timer(ELECTION_TIMEOUT_MAX), log(), commit_index(0),
+        state_machine(state_machine)
+    {
+        for (auto i = 0; i < peer_count; i++) {
+            peers[i]->match_index = 0;
+        }
+    }
+
+    bool process(const Message& msg, Message &reply)
+    {
+        reply.from_id = id;
+
+        switch (msg.type) {
+            case Message::Type::VoteRequest: {
+                DEBUG("Got a VoteRequest from %d", msg.from_id);
+                reply.type = Message::Type::VoteReply;
+                reply.vote_reply.vote_granted = false;
+
+                const auto req_lli = msg.vote_request.last_log_index;
+                const auto req_llt = msg.vote_request.last_log_term;
+
+                if ((msg.term > term && log.last_index() <= req_lli && log.last_term() <= req_llt)
+                    ||
+                    (msg.term == term &&
+                     msg.from_id == voted_for)) {
+
+                    reply.vote_reply.vote_granted = true;
+                    term = msg.term;
+                    voted_for = msg.from_id;
+                    node_state = NodeState::Follower;
+
+                    DEBUG("Granted my vote to %d which has term %d", voted_for, term);
+                }
+
+                reply.term = msg.term;
+                return true;
+            }
+
+            case Message::Type::VoteReply: {
+                DEBUG("Got a VoteReply(granted = %d)", msg.vote_reply.vote_granted);
+                // TODO: How about potential duplicate votes =
+
+                if (node_state != NodeState::Candidate) {
+                    break;
+                }
+
+                vote_count ++;
```

Comment:
Shouldn't you check first `if (msg.vote_reply.vote_granted)` before counting this vote?